### ISSUE TITLE
Upstream to alpine linux

### ISSUE THREAD

**Issue by mitchty** [mitchty/alpine-linux-ghc-bootstrap#6]
Making this an issue to compel me to get this upstream by hook or by crook.

**Comment by mitchty** [comment]
Upstream patch submitted is here:
http://sprunge.us/CPAa

Email send to alpine-devel ML, will wait on status.

**Comment by mitchty** [comment]
Email thread:
http://lists.alpinelinux.org/alpine-devel/5135.html

**Comment by mitchty** [comment]
It was suggested to move the bootstrap stuff into a snapshot function in the apkbuild.

http://wiki.alpinelinux.org/wiki/APKBUILD_examples:Git_checkout

I think I'll just leave everything in the makefile for now and just do a make && scp. Also need to update the build process to then not use a BOOTSTRAP environment variable and detect if ghc is available at the start of the apkbuild process.

**Comment by mitchty** [comment]
Trying this upstreaming thing again:
http://patchwork.alpinelinux.org/patch/1682/
http://patchwork.alpinelinux.org/patch/1683/
http://patchwork.alpinelinux.org/patch/1681/

Will upstream cabal and stack as well once those are in place.

**Comment by mitchty** [comment]
Right so the plan now talking to the Alpine Linux guys is to remove the "special" ghc llvm port I have for llvm 3.7.1 to make a more generic llvm3.7 port that can be used by more than just ghc.

I have this working in the sense that it builds ghc 8.0 now again. the llvm3.7 port I have though hypothetically could be linked to by things like julia which was the other language I found out has similar llvm requirements.

The llvm 3.7 I have I built to be able to be installed and used alongside the system llvm so that should help things out.

Right now I'm rewriting all of my port makefiles to be a bit less hacky and setting up Jenkins and an arm build box to start doing builds more often and to be able to test builds via git checkouts.

End goal is to have something similar to hvr's apt repo so you can have multiple versions of ghc/stack/cabal installed to test things.

But I've about 5 hours to wait for things to finish compiling before I can commit anything llvm3.7 related.

**Comment by mitchty** [comment]
Latest upstreamed patch series based off this:
https://lists.alpinelinux.org/alpine-aports/3248.html
llvm3.7 http://patchwork.alpinelinux.org/patch/2235/
ghc-bootstrap http://patchwork.alpinelinux.org/patch/2236/
ghc http://patchwork.alpinelinux.org/patch/2237/
cabal http://patchwork.alpinelinux.org/patch/2238/
stack http://patchwork.alpinelinux.org/patch/2239/

Will see if this latest series gets any more traction than the last two.

**Comment by mitchty** [comment]
llvm3.7 got merged at least, progress. >.<

**Comment by mitchty** [comment]
gcc-bootstrap is upstreamed-ish to x86_64, armhf is blocked on other builds but so far building against that upstreamed gcc-bootstrap is working.

**Comment by mitchty** [comment]
Right so everything builds even armhf. Final PR is here https://github.com/alpinelinux/aports/pull/205

And then this repo can be a historical artifact and all future work should end up there.

**Comment by ChristopherBiscardi** [comment]
The progress here is awesome. If ghc, etc, gets upstreamed we'd likely be able to use it for some [alpine-based official images](https://github.com/freebroccolo/docker-haskell/issues/49) much more easily.

**Comment by mitchty** [comment]
Yep thats the goal, haven't had a ton of time lately to pursue it again after I hit the fun of text segments in shared libraries that shouldn't be there but I'll try and have a go at it this weekend. I have that bit sorted in the bootstrap bit, but need to figure out what causes ghci to fail to work after building that way. That or find a way to have the initial ghc configure ensure -fPIC/-DPIC are specified for the things its calling in configure itself (where the true issue lies I think)

**Comment by mitchty** [comment]
Never thought this would take over a year, but 8.0.2 is what the upstreamed version is. Guess I'll pester people a bit more often.

**Comment by mitchty** [comment]
And its upstreamed, armhf is missing and I'll likely have to redo the cross compilation nonsense. But... closing this issue and likely this repo.
